Initial input component scaffold

macandcheese opened the pull request:
First pass of a `<calcite-input>` component, still super rough - no rtl, tests, etc...

Opening as draft to get some eyes on it while I work though some cleanup, styling, config, certain interactions between native + custom label / input elements (has to do with "where to set the id") etc.

I was unable to get `shadow: true` to work with password managers - they can’t “see” inside the shadow DOM to determine where to include their custom elements. As a result there are no slots within `<calcite-input>`, and styling is not applied on :host but rather on `<calcite-input>` (this felt weird to me, but, I was unable to get `shadow:true` to work).

Of note (I thought this was specific to calcite-input but it actually occurs on native input as well) - password managers shred the DOM to re-create their "version" of an input with custom content added - as such focus event is emitted on first focus, then a blur event is emitted while the field is removed and replaced with custom DOM created by password manager, then another focus on the target input is emitted as the password manager returns focus to the intended input. This was at least my experience with 1password - in practice that means if you use a "focus" event to show an input message and a "blur" event to hide it, you'll get a "double flicker" as the pw manager blurs and re-focuses the input... anyway...

I included a super basic validation example on `index.html` to show how the three components can work together - pretty poor examples of "validation functions" right now but just wanted to show example of usage in "real" form example...

Still need to work through better custom number buttons, replacing the file input with something more custom, etc.

Encouraged usage:
```
<calcite-label status=“invalid”>
    My great label
    <calcite-input placeholder=“Enter your information”></calcite-input>
    <calcite-input-message active>Here’s something you should know about this input</calcite-input-message>
</calcite-label>
```

This perhaps means we can look at refactoring checkbox, switch, radio button to use this new calcite-label component (possibly with an "inline" display option for using next to those items vs. on top of?)

The storybook has some very basic examples, and I've added some more detailed docs for events / attributes to the readme for each of calcite-input, calcite-label, calcite-input-message - feel free to give feedback on the content of those as well.

calcite-input draft readme: https://github.com/Esri/calcite-components/blob/a7e5c80f2f9f78a60ef329960086acfcdf17738a/src/components/calcite-input/readme.md

calcite-input-message draft readme: https://github.com/Esri/calcite-components/blob/a7e5c80f2f9f78a60ef329960086acfcdf17738a/src/components/calcite-input-message/readme.md

calcite-label draft readme: https://github.com/Esri/calcite-components/blob/a7e5c80f2f9f78a60ef329960086acfcdf17738a/src/components/calcite-label/readme.md

macandcheese commented:
@Esri/calcite-components any thoughts on this?

macandcheese commented:
@Esri/calcite-components anyone want to give a look? Don't want to go too far down this path for API / structure before I get some more 👀on it since inputs will be pretty heavily used I'd imagine...

pr3tori4n commented:
@macandcheese the lack of password manager support is a really interesting constraint. Have you tested this with any other managers? It'd be worthwhile to know if this is a universal problem or if it's specific to one or two managers. I'm curious as to whether Chrome's built-in password manager is able to work with web components.

macandcheese commented:
To clarify, it didn't work with shadow "enabled" - this take should work, although I only tried with 1password and native browser password manager.

The "double event" thing I mention happens on native inputs as well with password managers as the original form is focused, blurred when the pw manager recreates dom, then the "new" version with pw manager dom added is focused (I believe anyway).

pr3tori4n commented:
@macandcheese I'm considering the pros and cons of this functionality being encapsulated in a single component or split across three. It seems to me like these might be better if rolled into a single component. I think it would simplify the implementation to combine the input + label into a single custom element. The status messages may make sense as a slot, especially if you're passing an unknown number in, like a valid/invalid and a custom icon for it, but what do you think about combining the input and the label?

macandcheese commented:
That's reasonable too... The biggest reason I didn't go that route to start was to keep these as composable as possible - allowing the combination of a `<calcite-label>` with a `<calcite-checkbox>` or `<calcite-switch>` -  or for whatever reason using a native input / native label with calcite input / calcite label. Open to adjusting though if others want it all in one.

pr3tori4n commented:
https://discussions.agilebits.com/discussion/79137/shadow-dom-polymer-forms-do-not-fill-companion-extension-v4-7-4-now-supports-shadow-doms
This might also be worth noting. It seems 1Password is currently working on supporting Web Components and has a beta out for testing.

macandcheese commented:
Nice - that'd be nice if those updates got rolled out - if feels weird to have input be the only "shadow:false" component (and lose ability to style on :host / attributes), but for now it does seem to work OOTB with all password managers 🤷‍♂.

macandcheese commented:
Pushed some more updates for feedback, still needs some styling work, test, RTL, etc.

macandcheese commented:
Added some more features and examples...

- prefix / suffix string support
- resize icon for textarea
- update to calcite-icon
- slotted input action button

... and started work on "inline labels" for use with calcite-checkbox, calcite-switch, etc. (not functional currently, just scaffolded out). Still need a lot of cleanup especially around file input, number type logic, etc.

macandcheese commented:
@asangma any style feedback or use cases not covered here for you all?

julio8a commented:
@jcfranco, @driskull, @paulcpederson, @vcolavin, @pr3tori4n, @patrickarlt 

Any feedback on this PR?

paulcpederson commented:
I like it!

jcfranco commented on src/components/calcite-input/calcite-input.tsx:
```diff
@@ -0,0 +1,366 @@
+  }
+
+  componentWillLoad() {
+    this.inputType = this.el.getAttribute("type");
```
Could this be done via properties and not attributes? Reasoning is that Stencil components focus more on props than attributes. Also, by not having props, the compiler won't be able to generate proper types.

macandcheese replied:
Yep good catch.

macandcheese replied:
Adding all "reasonable" options as allowable types (ie we probably don't need to allow checkbox, or radio, or range, since other components handle those (or, will in the case of radio)).

jcfranco commented on src/components/calcite-input/calcite-input.tsx:
```diff
@@ -0,0 +1,366 @@
+    const numberButtonsHorizontalDown = (
+      <div
+        class={`calcite-input-number-button-item ${numberButtonClassModifier}`}
+        onClick={e => this.handleNumberButtonClick(e)}
```
I believe this will generate a new function on each render. You could avoid it by rewriting the following way:

```tsx
 <div {/*...*/}
      onClick={this.handleNumberButtonClick}
      {/*...*/}
 >
 </div>

// ... 

private handleNumberButtonClick = (e) => {
  // ...
}
```

macandcheese replied:
Yes, good catch ⚾️

jcfranco commented on src/components/calcite-input/calcite-input.tsx:
```diff
@@ -0,0 +1,366 @@
+
+  //--------------------------------------------------------------------------
+  //
+  //  Public Methods
```
This comment header could be removed.

macandcheese commented on src/components/calcite-input/readme.md:
```diff
@@ -0,0 +1,173 @@
+
+`icon` = [`path string`] - example: `<calcite-input icon={chevronUp24}>`
+
+`prefix-string` and `suffix-string` allow you position strings in a leading and trailing position.
```
I was getting a reserved warning with prefix: `The @Prop() name "prefix" is a reserved public name.`

Maybe `prefix-text/suffix-text` is better than `-string`?

macandcheese replied:
Updated to `-text`. Open to suggestions for other names.

macandcheese commented on src/assets/demo/validateInput.js:
```diff
@@ -0,0 +1,135 @@
+
+function logBlur(event) {
+  let targetNotice = document.querySelector(
+    `#validate-input-example-blur-notice`
```
Yeah, this was just an example of how someone could make their own validation with these components, it's ... very bad lol, should probably just make a new example :)

macandcheese commented on src/components/calcite-input-message/calcite-input-message.scss:
```diff
@@ -0,0 +1,112 @@
+}
+
+//todo rtl
+.calcite-input-message-icon {
```
This component actually has `shadow: false` because shadow DOM isn't parsable by password managers, so using `:host` here doesn't work to style off of.

macandcheese commented on src/components/calcite-input/calcite-input.scss:
```diff
@@ -1,373 +1,335 @@
-:host {
+calcite-input {
```
Using `shadow: false` - so `:host` here doesn't work to style off of unfortunately.

macandcheese commented on src/components/calcite-label/calcite-label.tsx:
```diff
@@ -0,0 +1,140 @@
+
+  render() {
+    const theme = getElementTheme(this.el);
+    const attributes = this.getAttributes();
```
Yes, but it automatically spreads to the parent`calcite-component`- where we really want them is spread to the rendered `input` or `textarea` (or, in other components, some other element maybe).

So this passes any attributes directly to the rendered child element (and filters out props, which are just clutter on a non-Host element). This way users can add placeholder, autocomplete, validation patterns, etc. and they will be spread to the rendered child. 

Otherwise we'd need to explicitly define 35/40 possible input attributes (and still explicitly set them on the rendered child). Let me know if I'm missing something here.

macandcheese commented on src/assets/demo/validateInput.js:
```diff
@@ -0,0 +1,135 @@
+  } else existingMessage.active = true;
+}
+function validatePasswordExampleBlurMessage(event) {
+  let existingMessage = document.querySelector(`#pw-status-1`);
```
Yeah poor naming, this is just populating the "example message" with info as example of emits. Not sure the value of that, but it's just for the demo. The only fields I added example validation for were the password fields that update based on a simple length check.

macandcheese commented on src/assets/demo/validateInput.js:
```diff
@@ -0,0 +1,135 @@
+  } else existingMessage.active = true;
+}
+function validatePasswordExampleBlurMessageNative(event) {
+  let existingMessage = document.querySelector(`#pw-status-2`);
```
Yeah poor naming, this is just populating the "example message" with info as example of emits. Not sure the value of that, but it's just for the demo. The only fields I added example validation for were the password fields that update based on a simple length check.

macandcheese commented on src/components/calcite-input-message/calcite-input-message.scss:
```diff
@@ -0,0 +1,112 @@
+
+  :host([status="#{$name}"]) {
+    & .calcite-input-message-icon {
+      color: $color !important;
```
Yeah not needed here.

macandcheese commented on src/components/calcite-input-message/calcite-input-message.tsx:
```diff
@@ -0,0 +1,100 @@
+  //
+  //--------------------------------------------------------------------------
+
+  private setIcon(path) {
```
Yep, remnant from our simple lives before `<calcite-icon>` 😆

macandcheese commented on src/components/calcite-input/readme.md:
````diff
@@ -0,0 +1,173 @@
+</calcite-label>
+```
+
+### With native elements
````
Yeah good point, I'll remove the doc and just make sure some basic combinations work like in the examples in the case it does happen or an app has some specific need to do so.

macandcheese commented on src/components/calcite-label/calcite-label.tsx:
```diff
@@ -0,0 +1,140 @@
+    this.focusChildEl();
+  }
+
+  @Listen("keydown") handleKeydown() {
```
Ha, yeah this isn't needed at all in fact.

macandcheese commented on src/components/calcite-input-message/calcite-input-message.tsx:
```diff
@@ -0,0 +1,100 @@
+  @Prop({ mutable: true, reflect: true }) type: "default" | "floating" =
+    "default";
+
+  /** optionally pass icon path data - pass only raw path data from calcite ui helper  */
```
Considering removing the custom icon, for now, until the icon team updates their library to remove the "filled" designation. Reason being, any wrapping component currently needs to have extra props just to populate that on the rendered child, it feels weird to have "filledIcon" as a prop on a parent. They are supposedly removing "filled" and instead, keeping a few filled icons that are designated with "-filled" in the string.

julio8a commented:
Is this guy ready to merge?

jcfranco commented:
> Is this guy ready to merge?

Can we hold off until beta 5 ships?

macandcheese commented:
I would like to get some tests fixed and some things working in IE. Would like some more eyes on it TBH, it's a big PR and kind of an oddity since there's no shadow root used (our only component like this) and includes three components. 

Still outstanding in my mind:
- Inputs themselves need more tests written
- I am hoping to Improve the way we allow people to wrap slotted label content around a control ( think "on" - switch - "off"). It's challenging to parse two bits of text in a single unnamed slot, I'm close-ish but I think that's the most "native" experience so I'd like to ship with that
- general style cleanup needed

macandcheese commented:
@Esri/calcite-components - I've updated how the `inline` and `inline-space-between` layout options work, you should be able to position text and controls within a `calcite-label` and it will smartly apply spacing to be able to render a number of "toggle row" type control layouts - cc @asangma 

I've added a bunch of tests, everything is working down to Edge - predictably I am still fighting an IE11 style and render bug, then this should be good to go.

Once this is merged in we can remove the label facility we have in some of the form control components in favor of `calcite-label` - and update `calcite-date` and `calcite-combobox` to use the `calcite-input`. 

In Edge - while wrapping a `calcite-switch` - I am seeing https://github.com/Esri/calcite-components/issues/283, so that seems to be something we should address in that component.

julio8a commented:
🎉🎉🎉🎉🎉

bstifle commented:
woooooot!
